samples/host_exerciser/host_exerciser_cmd.h

```diff
@@ -36,26 +36,48 @@ namespace host_exerciser {
 
 class fpgaperf {
 public:
-  fpga_result perfget(token::ptr_t token, fpga_perf_counter *fpga_perf)
+  static std::shared_ptr<fpgaperf> get(token::ptr_t token)
   {
-      return fpgaPerfCounterGet(token->c_type(), fpga_perf);
+    /*std::make_shared function doesn’t have access to the private constructor
+     *of fpgaperf, and as such it cannot instantiate it. Created temporary
+     *class that inherits from fpgaperf.*/
+    struct shared_enabler : public fpgaperf {};
+    auto p = std::make_shared<shared_enabler>();
+    if (fpgaPerfCounterGet(token->c_type(), p->counter_) != FPGA_OK) {
+        std::cout << "Failed to get the fpga perf counter" << std::endl;
+        p.reset();
+        return NULL;
+    }
+    return p;
   }
-  fpga_result perfstart(fpga_perf_counter *fpga_perf)
+  ~fpgaperf()
   {
-      return fpgaPerfCounterStartRecord(fpga_perf);
+    if (fpgaPerfCounterDestroy(counter_) != FPGA_OK) {
+        std::cout << "Failed to destroy the fpga perf counter" << std::endl;
+    }
+    if(counter_) {
+        delete counter_;
+        counter_ = nullptr;
+    }
   }
-  fpga_result perfstop(fpga_perf_counter *fpga_perf)
+  fpga_result start()
   {
-      return fpgaPerfCounterStopRecord(fpga_perf);
+      return fpgaPerfCounterStartRecord(counter_);
   }
-  fpga_result perfprint(fpga_perf_counter *fpga_perf)
+  fpga_result stop()
   {
-      return fpgaPerfCounterPrint(stdout, fpga_perf);
+      return fpgaPerfCounterStopRecord(counter_);
   }
-  fpga_result perfdestroy(fpga_perf_counter *fpga_perf)
+  fpga_result print()
```

Comment:
When moving to c++, try to simplify the code.
For example, print seems like it should return a void (unless you thing there's value in capturing the error code).
If you happen to catch errors while performing some lower level operation, then you can print the message here.
Also, you might want to switch to using c++ exceptions for error notification and handling instead of error codes.